source/fuzz/fuzzer_pass.cpp

```diff
@@ -27,5 +27,97 @@ FuzzerPass::FuzzerPass(opt::IRContext* ir_context, FactManager* fact_manager,
+      result.push_back(&*prev_inst_it);
+    }
+  }
+
+  // Walk the dominator tree to consider all instructions from dominating
+  // blocks
+  auto dominator_analysis = GetIRContext()->GetDominatorAnalysis(&function);
+  for (auto next_dominator = dominator_analysis->ImmediateDominator(block);
+       next_dominator != nullptr;
+       next_dominator =
+           dominator_analysis->ImmediateDominator(next_dominator)) {
+    for (auto& dominating_inst : *next_dominator) {
+      if (instruction_is_relevant(GetIRContext(), &dominating_inst)) {
+        result.push_back(&dominating_inst);
+      }
+    }
+  }
+  return result;
+}
+
+void FuzzerPass::MaybeAddTransformationBeforeEachInstruction(
+    std::function<uint32_t(
+        const opt::Function& function, opt::BasicBlock* block,
+        opt::BasicBlock::iterator inst_it, uint32_t base, uint32_t offset)>
+        maybe_apply_transformation) {
+  // Consider every block in every function.
+  for (auto& function : *GetIRContext()->module()) {
+    for (auto& block : function) {
+      // We now consider every instruction in the block, randomly deciding
+      // whether to apply a transformation before it.
+
+      // In order for transformations to insert new instructions, they need to
+      // be able to identify the instruction to insert before.  We enable this
+      // by tracking a base instruction, which must generate a result id, and
+      // an offset (to allow us to identify instructions that do not generate
+      // result ids).
+
+      // The initial base instruction is the block label.
+      uint32_t base = block.id();
+      uint32_t offset = 0;
+      // Consider every instruction in the block.
+      for (auto inst_it = block.begin(); inst_it != block.end(); ++inst_it) {
+        if (inst_it->HasResultId()) {
+          // In the case that the instruction has a result id, we use the
+          // instruction as its own base, with zero offset.
+          base = inst_it->result_id();
+          offset = 0;
+        } else {
+          // The instruction does not have a result id, so we need to identify
+          // it via the latest instruction that did have a result id (base), and
+          // an incremented offset.
+          offset++;
+        }
+
+        // Invoke the provided function, which might apply a transformation.
+        // Its return value informs us of how many instructions it inserted.
+        // (This will be 0 if no transformation was applied.)
+        uint32_t num_instructions_inserted =
+            maybe_apply_transformation(function, &block, inst_it, base, offset);
+
+        if (!inst_it->HasResultId()) {
+          // We are tracking the current id-less instruction via an offset,
+          // |offset|, from a previous instruction, |base|, that has an id. We
+          // increment |offset| to reflect any newly-inserted instructions.
+          offset += num_instructions_inserted;
```

Comment:
Could add a comment explaining that, even though the inserted instructions might include one with a result id, you are not going to use that result id.

Reply:
Done.